### ISSUE TITLE
False-negative: Method names with underscores

### ISSUE THREAD

**Issue by mohitsatr** [checkstyle/checkstyle#17841]
From: https://google.github.io/styleguide/javaguide.html#s5.3-camel-case
> In very rare circumstances (for example, multipart version numbers), you may need to use underscores to separate adjacent numbers, since numbers do not have upper and lower case variants.

Google style is not exact. On one hand it says `_` can be used in test methods to combine sentences and on another hand in general CamelCase explanation, `_` is allowed for numbering suffix in all names(except package and generic names).

Valid uses of underscore :
```
1. guava33_4_5
2. guavaVersion33_4_5 
3. guavaversion33_4_5 (same as 1.)
```

Invalid uses of underscore:
```
1. guava_33_4_5_ (underscore at the end)
2. guava_33_4_5   (underscore between digit and letter)
3. guava_version33_4_5 (underscore between lowercase character sequences)
```

**Note**: Underscores are not allowed in normal method names. Name has to have the numbering format part at the end. Following are all invalid

```
guava_version
guava_Version
Set_guavaVersion3345
```

Some false-negatives left to address from #17708  

```java
$ cat NamesWithUnderscores.java
/** some javadoc. */
public class NamesWithUnderscores {
  void gradle9_5_1_() {} // Expected violation, _ at the end

  void gradle_9_5_1() {} // Expected violation, _ between digit and letter

  void jdk_9_0_392() {} // Expected violation, _ between digit and letter

  void testing_01231(String str) {} // Expected violation, _ between digit and letter
}
```

```
$ java -jar checkstyle-11.1.0-all.jar -c google_checks.xml NamesWithUnderscores.java 
Starting audit...
[WARN] /mnt/5D92528E6B945467/test/testing/NamesWithUnderscores.java:4:8: Method name 'gradle9_5_1_' must match pattern '^(?![a-z]$)(?![a-z][A-Z])[a-z][a-z0-9]*(?:[A-Z][a-z0-9]*)*(?:_[0-9]+)*$'. [MethodName]
Audit done.
```

We probably need a better approach than Regex. Current regex has already grown very complex.
 
https://github.com/checkstyle/checkstyle/blob/5d52248c85b0fc37ddd786ddb5354e332f8a54b8/src/main/resources/google_checks.xml#L436-L438

if special check is needed, it should first determine the type of name - regular or name with numbering and then it should validate the name against the appropriate regex format, which will differ for each type of naming

Keep in mind: unless it's a non-regular name i.e name that has number formating suffix at the end or method with `@Test` annotation,  underscores are not allowed.

**Comment by romani** [comment]
current Check is https://checkstyle.sourceforge.io/checks/naming/methodname.html#MethodName

we can make `GoogleMethodName` that has no properties and does exactly what Google need, even treatment of `Test` annotation can be hardcoded.

**Comment by aryanttripathi** [comment]
Hey @mohitsatr @romani Can i work on this issue? I am new to checkstyle and want to contribute in it. Also can you explain a little what changes actually need to made, do we need to break on the basis of conditions?

**Comment by mohitsatr** [comment]
@Aryant-Tripathi Please start with [good xxx issues](https://github.com/checkstyle/checkstyle/issues?q=is%3Aissue%20state%3Aopen%20label%3A%22good%20second%20issue%22) first.

Please watch set-up [tutorials](https://www.youtube.com/playlist?list=PLHM9s_lN4X0hzOQ0sUmGdroxW0HfREAqj)

**Comment by vivek-0509** [comment]
I am on it

**Comment by Zopsss** [comment]
@mohitsatr how is this valid?

```java
void gradle9_5_1_() {} // ok
```

This is a similar case to:

```
Invalid uses of underscore:

1. guava_33_4_5_ (underscore at the end)
```

**Comment by Zopsss** [comment]
@vivek-0509 Please create a separate issue for this new check, we first need a solid understanding of what all the things this new check should cover. The issue description should cover all the things we can expect from this new check and the reason behind creating it.

**Comment by mohitsatr** [comment]
@Zopsss, Yes you are right. Fixed

**Comment by vivek-0509** [comment]
@

> [@vivek-0509](https://github.com/vivek-0509) Please create a separate issue for this new check, we first need a solid understanding of what all the things this new check should cover. The issue description should cover all the things we can expect from this new check and the reason behind creating it.

Sure

**Comment by romani** [comment]
new Check is merged and placed to google style config.
